### PR TITLE
add API to get compatible modules

### PR DESCRIPTION
**Pull request by karthikprasad**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs change / refactoring / dependency upgrade

## Motivation and Context / Related issue
From @ffuuugor:
> There could be an issue with doing model fixing and module replacement in make_private.
Imagine scenario (not a hypothetical, I'm fixing dcgan example now):
> 1) you create a model and an optimizer. And you pass model.parameters() to the optimizer, because that's how you initialise optimizer
> 2) you then pass aforementioned model and optimizer to make_private() with try_fix_incompatible=True. It does clone_module for the root and a bunch of children and returns the result.
> 3) Now we have a problem: optimizer is pointing to the parameters of the old model before fixing, but gradients are computed on the parameters of the new model, after fixing

We can't simply clone the optimizer and supply new model's parameters either (see #225)

## How Has This Been Tested (if it applies)

unittest passes on top of ffuuugor_experimental_merge

## Checklist

- [ ] The documentation is up-to-date with the changes I made.
- [x] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by romovpa** [comment]
@karthikprasad Check plz if I understood the solution correctly. 

If the model is not valid, we suggest user to apply the transform and fix it explicitly:
```
model = models.resnet18()
fixed_model = privacy_engine.get_compatible_module(model)
```

After this, `model` and `fixed_model` are two independent models with their own physical parameters. As a common case we should use the latter for further computations.

File: opacus/privacy_engine.py [hunk 1]
```diff
@@ -207,6 +203,9 @@ def _prepare_model(
         try_fix_incompatible_modules: bool = False,
```
Comment (ffuuugor) [inline opacus/privacy_engine.py:203]: do we want to remove `try_fix_incompatible_modules` from here as well?

File: opacus/tests/privacy_engine_test.py [hunk 2]
```diff
@@ -102,25 +102,25 @@ def _init_private_training(
         noise_multiplier: float = 1.0,
         max_grad_norm: float = 1.0,
         poisson_sampling: bool = True,
-        try_fix_incompatible_modules: bool = False,
     ):
+        privacy_engine = PrivacyEngine(secure_mode=secure_mode)
+
         model = SampleConvNet()
+        model = privacy_engine.get_compatible_module(model)
```
Comment (ffuuugor) [inline opacus/tests/privacy_engine_test.py:109]: I'd rather have a test both both occasions: where we call `get_compatible_module` and where we don't.
Both should work (obvs provided the model is compatible in the first place). 

The original bug was spotted by dumb luck: `dcgan` example just turned out to be the only place where we used fixing behaviour, but it wasn't designed for it.
Let's make sure we have dedicated test for this

**Review by ffuuugor — CHANGES_REQUESTED** [review]
Thanks for a prompt and solid fix @karthikprasad  !
(No, really, Karthik submitted the diff less than 30 minutes after a bug was discovered. We still move fast sometimes)

Few inline comments

File: opacus/privacy_engine.py [hunk 3]
```diff
@@ -56,6 +56,15 @@ def validate(
         """
         ModuleValidator.validate(module, raise_if_error=True)
 
+    def get_compatible_module(self, module: nn.Module) -> nn.Module:
```
[the remark this answers is not in the document]
Comment (karthikprasad) [reply opacus/privacy_engine.py:59]: >We can make it a @classmethod.

I don't see why not. Makes more sense since it doesn't depend on any instance attributes.

>what's the downside of asking people to call ModelValidator.fix() directly?

1. Users having to import `ModuleValidator` in their training file (in addition to PrivacyEngine).
2. We lose the [single-window system ](https://www.wikiwand.com/en/Single-window_system) for accessing Opacus internals.

**Comment by karthikprasad** [comment]
> @karthikprasad Check plz if I understood the solution correctly.
> 
> If the model is not valid, we suggest user to apply the transform and fix it explicitly:
> 
> ```
> model = models.resnet18()
> fixed_model = privacy_engine.get_compatible_module(model)
> ```
> 
> After this, `model` and `fixed_model` are two independent models with their own physical parameters. As a common case we should use the latter for further computations.

@romovpa That is correct!

File: opacus/privacy_engine.py (hunk 1 above)
Comment (karthikprasad) [reply opacus/privacy_engine.py:203]: I kept it as is since, technically the user can call `privacy_engine._prepare_model()` with the flag. However, the method is meant to be private, so it doesn't make sense anyway. I'll remove it

File: opacus/privacy_engine.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ffuuugor) [reply opacus/privacy_engine.py]: >Users having to import ModuleValidator in their training file

That make sense. I don't have a strong preference, and this argument alone is enough to justify this decision I think.

> We lose the single-window system for accessing Opacus internals.

As a more general point, I'd love to have a discussion about this at some point. I would argue that having a single-window system is not the most fitting design approach, despite opacus historically being this way.

**Review by ffuuugor — APPROVED** [review]
LGTM, Thank you!

JFYI, I've merged the main branch, so now the only two tests that are expected to fail are multi_gpu and linter on vision_benchmark.
